pallets/process-validation/src/lib.rs

```diff
@@ -79,53 +90,110 @@ pub mod pallet {
+
+    #[pallet::storage]
+    #[pallet::getter(fn version_model)]
+    pub(super) type VersionModel<T: Config> = StorageMap<_, Blake2_128Concat, T::ProcessIdentifier, i32, ValueQuery>;
 
     #[pallet::event]
     #[pallet::generate_deposit(pub(super) fn deposit_event)]
     pub enum Event<T: Config> {
-        // TODO: implement correct events for extrinsics including params
-        ProcessCreated,
-        ProcessDisabled,
+        // id, version, restrictions, is_new
+        ProcessCreated(T::ProcessIdentifier, i32, Vec<Restriction>, bool),
+        //id, version, updated
+        ProcessDisabled(T::ProcessIdentifier, i32, bool),
     }
 
     #[pallet::error]
     pub enum Error<T> {
-        // TODO: implement errors for extrinsics
+        NonExistingProcess,
+        AlreadyDisabled,
+        InvalidVersion,
     }
 
     // The pallet's dispatchable functions.
     #[pallet::call]
     impl<T: Config> Pallet<T> {
-        // TODO: implement create_process with correct parameters and impl
         #[pallet::weight(T::WeightInfo::create_process())]
-        pub(super) fn create_process(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-            // Check it was signed and get the signer
+        pub(super) fn create_process(
+            origin: OriginFor<T>,
+            id: T::ProcessIdentifier,
+            restrictions: Vec<Restriction>,
+        ) -> DispatchResultWithPostInfo {
             T::CreateProcessOrigin::ensure_origin(origin)?;
+            let new_version: i32 = Pallet::<T>::_update_version(id.clone()).unwrap();
+            Pallet::<T>::_persist_process(&id, &new_version, restrictions.clone());
 
-            Self::deposit_event(Event::ProcessCreated);
-            Ok(().into())
+            // wrap event data into a struct?
+            Self::deposit_event(Event::ProcessCreated(id, new_version, restrictions, new_version == 1));
+
+            return Ok(().into());
         }
 
-        // TODO: implement disable_process with correct parameters and impl
         #[pallet::weight(T::WeightInfo::disable_process())]
-        pub(super) fn disable_process(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+        pub(super) fn disable_process(
+            origin: OriginFor<T>,
+            id: T::ProcessIdentifier,
+            version: i32,
+        ) -> DispatchResultWithPostInfo {
             T::DisableProcessOrigin::ensure_origin(origin)?;
-            Self::deposit_event(Event::ProcessDisabled);
-            Ok(().into())
+
+            // TODO implement errors
+            // TODO if no process for this version return latest along with the error message
+            if !<ProcessModel<T>>::contains_key(id.clone(), version) {
+                Self::deposit_event(Event::ProcessDisabled(id, version, false));
+                return Ok(().into());
+            }
+
+            let mut updated = false;
+
+            <ProcessModel<T>>::mutate(id.clone(), version, |process| {
```

Comment:
In answer to your slack question definitely mutate over get and insert. It's just more idiomatic given what we're doing

Reply:
:+1: this has been updated a little.